package/kedro_viz/server.py

```diff
@@ -198,29 +199,50 @@ def pretty_name(name):
     edges = []
     namespace_tags = defaultdict(set)
     all_tags = set()
+    namespace_to_layer = {}
+
+    data_set_to_layer_map = {
+        ds_name: getattr(ds_obj, "_layer", None)
+        for ds_name, ds_obj in catalog._data_sets.items()  # pylint: disable=protected-access
+    }
+
+    all_layers = sorted(set(layer for layer in data_set_to_layer_map.values() if layer))
```

Comment:
I don't think alphabetically sorting layers is the desired behaviour. I'm checking with Yetu and the viz team to see where we should encode the order information.

Reply:
@idanov suggested that we will have to deduce the layer's ordering based on the pipeline's topology. An idea I have is after parsing node's layer information, we use `pipeline.grouped_nodes` to deduce a sorted list of layers. If the order of dataset's layer can't be deduced, maybe we can omit it? Not sure... What do you think? @andrii-ivaniuk -- We can also bring it up at tech design tomorrow.

Reply:
This ordering doesn't have any effect. If you take a look at the example response you can see that we have `layers` list which is a mapping between layer name and some ID (in this case just sequence number). Then each node refers to layer by ID (like in RDBMS). I don't see the need to do something extra like deduce a sorted list of layers.

Reply:
We can consider it as a relational database with two tables: `layers` and `nodes`.
`node.layer` is a foreign key on `layer.id` and that's it.

Reply:
I was thinking more like how does the frontend know if they should display the `raw` layer first or the `final` layer first (visually) but maybe the frontend works it out automatically so this ordering is indeed not an issue. @richardwestenra could you give us an advice here please?

Reply:
I believe the front-end does rely on the ordering, yes. `layers[0]` is the top/first one.

Reply:
I mean, it calculates the positioning automatically, but having them in order is important for calculating the padding between layers. I could do it automatically but it'd be trickier, so ideally it'd be better to have them in order

Reply:
I feel like calculating and returning layers in order is the backend's responsibility. Otherwise, why bother returning the list of layers at all? Why can't we just return the layer's name inside the `node` and let the frontend deduce whatever information it needs to deduce?

Reply:
@limdauto We have an execution order defined by edges. Having this order we know the start point and its layer. The strat point layer is the first layer.

E.g. We have two pipelines with the following layer order:
1. raw -> null -> intermediate -> null -> final
2. new -> intermediate -> pre-final
Also, we have one dataset with `unused` layer in catalog.yml but it is not used in any pipeline

What should the final order be?

Reply:
I don't know. @richardwestenra what do you think the behaviour / backend response for the above example should be?

Reply:
- I would probably leave out any layer that isn't used by any of the nodes, but it doesn't really matter as it'll be filtered ultimately regardless.
- `null` isn't a layer, it's the absence of a layer, so it shouldn't be in the layer order object. so `raw -> null -> intermediate -> null -> final1` becomes `[raw, intermediate, final1]`

Reply:
What about `new` and `final2`? I think @andrii-ivaniuk meant those 2 pipelines are combined into one.

Reply:
@richardwestenra But we have two separate pipelines. 
Let's introduce some layer naming convention according to which each layer should be prefixed with a sequence number e.g. 1_raw, ..., 5_final.

Reply:
oh. Are we talking about modular/sub-pipelines? The current API architecture doesn't work for that anyway - I feel like it shouldn't be reflected in this PR, and we can cross that bridge when we add modular pipeline support in KedroViz in a subsequent PR?

btw, for context: The `name` field will be displayed to the user in Kedro-Viz, so it should be readable and user-friendly. The `id` field can be used for a UID if you need to avoid confusion/duplication. If you need to add extra context (e.g. to illustrate which sub-pipeline something belongs to) then maybe that could be added to a separate property on the object?

![image](https://user-images.githubusercontent.com/1155816/79250157-2fb73d00-7e76-11ea-9082-035df9b92cc9.png)

Reply:
I think in @andrii-ivaniuk's example, then maybe the final order should be `["new", "raw", "intermediate", "final1", "final2"]`. The first level of ordering is based on pipeline's topology. The second level of ordering (e.g. "new", "raw") can just be alphabetical. WDYT?

Reply:
@richardwestenra I didn't talk about the modular pipeline or I don't understand what do you mean. I talk about the following example:
<img width="977" alt="Screen Shot 2020-04-14 at 20 08 06" src="https://user-images.githubusercontent.com/24272541/79253410-d65a0880-7e8b-11ea-981c-8c37a2284411.png">


@limdauto Why did you make such an assumption? 
My thoughts: we have a lot of assumptions that result in a lot of edge cases.

Reply:
It wasn't really an assumption. It was more like an idea. We do know for a fact that we need to return the ordering of layers __somehow__ to the frontend. Part of this ordering is dictated by node's execution order (the pipeline's topology), so that part can't be changed. In your example, this part of ordering is `"new"` and `"raw"` needs to come before `"intermediate"` and `"intermediate"` needs to come before `"final1"` and `"final2"`. For the ordering that can't be inferred, it needs to be based on convention. Your initial implementation, ie alphabetical order, is as good a convention as any. If you have any other suggestion, would love to learn more.

I think what we should do is come up with a list of possible options here and bring it up with the larger group during tech design / post standup tomorrow.

Reply:
huh, ok. I mean, at the moment I think it assumes that you only have a single pipeline? I'm not sure the metaphor works with multiple unconnected pipelines displayed at once. hmm.

Reply:
I think this edge case needs product input. We will bring this up with the group tomorrow. To recap, the summary of the problem is: we are not sure where to put the logic to determine the ordering of layers. There are 3 places we can solve this problem:

* User: letting user specify the ordering of layers themselves (e.g. through a prefix convention or just stating them explicitly in catalog.yml)
* Backend: deducing the layers ordering based on node's execution order and some other criteria for the edge case presented above.
* Frontend: based on how best things look visually (?)

Let's bring it up after standup tomorrow @andrii-ivaniuk @richardwestenra